### DISCUSSION TITLE
Support media_player.turn_on / turn_off (e.g. Power On / Off) services / controls.

### DISCUSSION THREAD

**Discussion by thlucas1** [punxaphil/custom-sonos-card#338] [Ideas]
Any chance of adding standard `media_player.turn_on` / `turn_off` support for non-Sonos players?  The ability to turn on / off individual speakers (as well as a group of speakers at once) would be nice.  I know I can add a separate panel of buttons with tap actions to do this, but it would make for a nicer appearance if the card supported it natively.

I'm currently testing this with my Bose SoundTouch devices (don't hate me Sonos fans), and it's going well.  Just finding a few minor things like this.  If you're interested in supporting non-Sonos gear I would be happy to be a beta tester for you.  If not, I understand.

**Comment by punxaphil** [comment]
Can you share info about your devices? http://homeassistant.local/developer-tools/state

Also, I am working on https://github.com/punxaphil/maxi-media-player, a generic player that does not have a focus on Sonos players only. Feel free to try it out and help me improve it!

**Comment by thlucas1** [comment]
Thanks for the info Phil - I was not aware there was a generic player.  Looks promising.  I will look into it and touch base.

Here's more information on my GitHub SoundTouchPlus wiki page for my [SoundTouch 10](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus/wiki/SoundTouch-Model-Notes#soundtouch-10) and [SoundTouch 300](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus/wiki/SoundTouch-Model-Notes#soundtouch-300) devices.  They are basically speakers with a built in web server that hosts the [Bose SoundTouch Web Services API](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus/wiki/SoundTouch-WebServices-API), which provides endpoints to control / configure the speaker and interface with various music services.  I wrote the [Bose SoundTouch API Python3 Library](https://github.com/thlucas1/bosesoundtouchapi) to interface with those web services from Python, which is what I use from my [SoundTouchPlus integration](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus) to control the speaker(s).

I had already started looking into the Sonos Card, and have applied a couple of changes in development.  I corrected the issue of the player not muting or setting a volume level when it is the only player specified.  I was also in the process of adding power on / off button to the control deck area.

Thanks!

**Reply by punxaphil** [reply]
👍 
Please go into http://homeassistant.local/developer-tools/state and share a screenshot here of the info you get for your bose players

**Reply by thlucas1** [reply]
Bose SoundTouch 10 device (simple wireless speaker) - I have 4 of them, but they all contain the same values (with names changed).  The `group_members` indicates the ST-10 and ST-300 are synced and playing in a zone, with ST-10 the master.
```
source_list:
  - Airplay
  - Alexa
  - Aux In
  - Bluetooth
  - NAS Library THLUCASI9
  - Notification
  - Pandora (xxxx@xxxx.com)
  - Spotify (xxxx@xxxx.com)
  - THLUCASI9 Music
  - Tunein
  - Upnp
group_members:
  - media_player.bose_st10_1
  - media_player.bose_st300
soundtouchplus_nowplaying_isadvertisement: false
soundtouchplus_nowplaying_isfavorite: false
soundtouchplus_presets_lastupdated: 0
soundtouchplus_recents_lastupdated: 0
soundtouchplus_source: SPOTIFY:31l77y2al5lnn7xxxxxxxxx
soundtouchplus_sound_mode: not capable
soundtouchplus_tone_bass_level: not capable
soundtouchplus_tone_treble_level: not capable
device_class: speaker
icon: mdi:speaker
friendly_name: Bose-ST10-1
supported_features: 1040319
volume_level: 0.14
is_volume_muted: false
media_duration: 306
media_position: 144
media_position_updated_at: "2024-04-08T14:29:03.541581+00:00"
source: Spotify (xxxx@xxxx.com)
media_title: The Lion and the Lamb
media_artist: Big Daddy Weave
media_album_name: Beautiful Offerings (Deluxe Edition)
media_track: The Lion and the Lamb
shuffle: true
repeat: "off"
entity_picture: >-
  /api/media_player_proxy/media_player.bose_st10_1?token=307b0923e4d1618fbf6323af193f9fe7a3d2735d61d8fb0d6165ebe682ef8719&cache=bc23987020e538b9
```

SoundTouch 300 device (sound bar with bass module / more features).
Note the extra `sound_mode`, `sound_mode_list`, `soundtouchplus_x` attributes for sound mode feature support.
```
source_list:
  - Airplay
  - Bluetooth
  - Local Internet Radio
  - NAS Library THLUCASI9
  - Notification
  - Pandora (xxxx@xxxx.com)
  - Product (HDMI_1)
  - Product (TV)
  - Spotify (xxxx@xxxx.com)
  - THLUCASI9 Music
  - Tunein
  - Upnp
sound_mode_list:
  - Dialog
  - Normal
group_members:
  - media_player.bose_st300
soundtouchplus_nowplaying_isadvertisement: false
soundtouchplus_nowplaying_isfavorite: false
soundtouchplus_presets_lastupdated: 0
soundtouchplus_recents_lastupdated: 0
soundtouchplus_source: SPOTIFY:31l77y2al5lnn7xxxxxxxxx
soundtouchplus_sound_mode: AUDIO_MODE_DIRECT
soundtouchplus_tone_bass_level: 25
soundtouchplus_tone_bass_level_range: Min=-100, Max=100
soundtouchplus_tone_treble_level: 25
soundtouchplus_tone_treble_level_range: Min=-100, Max=100
device_class: speaker
icon: mdi:speaker
friendly_name: Bose-ST300
supported_features: 1040319
volume_level: 0.18
is_volume_muted: false
media_duration: 306
media_position: 270
media_position_updated_at: "2024-04-08T14:27:03.643605+00:00"
media_title: The Lion and the Lamb
media_artist: Big Daddy Weave
media_album_name: Beautiful Offerings (Deluxe Edition)
media_track: The Lion and the Lamb
source: Spotify (xxxx@xxx.com)
sound_mode: Normal
shuffle: true
repeat: "off"
entity_picture: >-
  /api/media_player_proxy/media_player.bose_st300?token=ea47fefb6838d31ff40ae09f9087d365cc47e3af060227aa521142f51a116ed4&cache=bc23987020e538b9
```

**Reply by punxaphil** [reply]
Please also go here: http://homeassistant.local/developer-tools/template
and paste:
```
{% for state in states.media_player -%}
  {{ state.attributes}}
{% endfor %}
```
and share the output here

**Reply by thlucas1** [reply]
```
{% for state in states.media_player -%}
  {{ state.attributes}}
{% endfor %}
```
Result type: string

```
{'source_list': ['Airplay', 'Alexa', 'Aux In', 'Bluetooth', 'NAS Library THLUCASI9', 'Notification', 'Pandora (xxxx@xxxx.com)', 'Spotify (xxxx@xxxx.com)', 'THLUCASI9 Music', 'Tunein', 'Upnp'], 'group_members': ['media_player.bose_st10_1', 'media_player.bose_st300'], 'volume_level': 0.19, 'is_volume_muted': False, 'media_duration': 206, 'media_position': 63, 'media_position_updated_at': datetime.datetime(2024, 4, 8, 15, 30, 33, 549006, tzinfo=datetime.timezone.utc), 'media_title': 'Prodigal', 'media_artist': 'Sidewalk Prophets', 'media_album_name': 'Something Different (Deluxe Version)', 'media_track': 'Prodigal', 'source': 'Spotify (xxxx@xxxx.com)', 'shuffle': True, 'repeat': 'off', 'soundtouchplus_nowplaying_isadvertisement': False, 'soundtouchplus_nowplaying_isfavorite': False, 'soundtouchplus_presets_lastupdated': 0, 'soundtouchplus_recents_lastupdated': 0, 'soundtouchplus_source': 'SPOTIFY:31l77y2al5lnn7xxxxxxxxx', 'soundtouchplus_sound_mode': 'not capable', 'soundtouchplus_tone_bass_level': 'not capable', 'soundtouchplus_tone_treble_level': 'not capable', 'device_class': 'speaker', 'entity_picture': '/api/media_player_proxy/media_player.bose_st10_1?token=307b0923e4d1618fbf6323af193f9fe7a3d2735d61d8fb0d6165ebe682ef8719&cache=abe9a83822fff743', 'icon': 'mdi:speaker', 'friendly_name': 'Bose-ST10-1', 'supported_features': <MediaPlayerEntityFeature.PAUSE|SEEK|VOLUME_SET|VOLUME_MUTE|PREVIOUS_TRACK|NEXT_TRACK|TURN_ON|TURN_OFF|PLAY_MEDIA|VOLUME_STEP|SELECT_SOURCE|STOP|PLAY|SHUFFLE_SET|SELECT_SOUND_MODE|BROWSE_MEDIA|REPEAT_SET|GROUPING: 1040319>}
{'source_list': ['Bose-ST10-1 Group', 'Web Player (Chrome)'], 'icon': 'mdi:spotify', 'friendly_name': 'SpotifyPlus Todd L', 'supported_features': <MediaPlayerEntityFeature.PAUSE|SEEK|VOLUME_SET|VOLUME_MUTE|PREVIOUS_TRACK|NEXT_TRACK|TURN_ON|TURN_OFF|PLAY_MEDIA|VOLUME_STEP|SELECT_SOURCE|PLAY|SHUFFLE_SET|BROWSE_MEDIA|REPEAT_SET|MEDIA_ENQUEUE: 2543551>}
{'source_list': ['Airplay', 'Bluetooth', 'Local Internet Radio', 'NAS Library THLUCASI9', 'Notification', 'Pandora (xxxx@xxxx.com)', 'Product (HDMI_1)', 'Product (TV)', 'Spotify (xxxx@xxxx.com)', 'THLUCASI9 Music', 'Tunein', 'Upnp'], 'sound_mode_list': ['Dialog', 'Normal'], 'group_members': ['media_player.bose_st300'], 'volume_level': 0.18, 'is_volume_muted': False, 'media_duration': 206, 'media_position': 63, 'media_position_updated_at': datetime.datetime(2024, 4, 8, 15, 30, 33, 671368, tzinfo=datetime.timezone.utc), 'media_title': 'Prodigal', 'media_artist': 'Sidewalk Prophets', 'media_album_name': 'Something Different (Deluxe Version)', 'media_track': 'Prodigal', 'source': 'Spotify (xxxx@xxxx.com)', 'sound_mode': 'Normal', 'shuffle': True, 'repeat': 'off', 'soundtouchplus_nowplaying_isadvertisement': False, 'soundtouchplus_nowplaying_isfavorite': False, 'soundtouchplus_presets_lastupdated': 0, 'soundtouchplus_recents_lastupdated': 0, 'soundtouchplus_source': 'SPOTIFY:31l77y2al5lnn7xxxxxxxxx', 'soundtouchplus_sound_mode': 'AUDIO_MODE_DIRECT', 'soundtouchplus_tone_bass_level': 25, 'soundtouchplus_tone_bass_level_range': 'Min=-100, Max=100', 'soundtouchplus_tone_treble_level': 25, 'soundtouchplus_tone_treble_level_range': 'Min=-100, Max=100', 'device_class': 'speaker', 'entity_picture': '/api/media_player_proxy/media_player.bose_st300?token=ea47fefb6838d31ff40ae09f9087d365cc47e3af060227aa521142f51a116ed4&cache=abe9a83822fff743', 'icon': 'mdi:speaker', 'friendly_name': 'Bose-ST300', 'supported_features': <MediaPlayerEntityFeature.PAUSE|SEEK|VOLUME_SET|VOLUME_MUTE|PREVIOUS_TRACK|NEXT_TRACK|TURN_ON|TURN_OFF|PLAY_MEDIA|VOLUME_STEP|SELECT_SOURCE|STOP|PLAY|SHUFFLE_SET|SELECT_SOUND_MODE|BROWSE_MEDIA|REPEAT_SET|GROUPING: 1040319>}
{'source_list': ['Bose-ST10-1 Group', 'Web Player (Chrome)'], 'volume_level': 0.18, 'media_content_id': 'spotify:track:041qpPpNFBg7UFvQFD1G9R', 'media_content_type': <MediaType.MUSIC: 'music'>, 'media_duration': 206.401, 'media_position': 85.929, 'media_position_updated_at': datetime.datetime(2024, 4, 8, 15, 30, 56, 621810, tzinfo=datetime.timezone.utc), 'media_title': 'Prodigal', 'media_artist': 'Sidewalk Prophets', 'media_album_name': 'Something Different (Deluxe Version)', 'media_track': 1, 'media_playlist': 'Worship Playlist', 'source': 'Bose-ST10-1 Group', 'shuffle': True, 'repeat': <RepeatMode.OFF: 'off'>, 'entity_picture': '/api/media_player_proxy/media_player.spotify_todd_l?token=bd4ed65591d73a164c01481e1947d0b536aaad98c75e115f05346357e013a15d&cache=abe9a83822fff743', 'friendly_name': 'Spotify Todd L', 'supported_features': <MediaPlayerEntityFeature.PAUSE|SEEK|VOLUME_SET|PREVIOUS_TRACK|NEXT_TRACK|PLAY_MEDIA|SELECT_SOURCE|PLAY|SHUFFLE_SET|BROWSE_MEDIA|REPEAT_SET: 444983>}
{'restored': True, 'supported_features': 152463}
{'restored': True, 'supported_features': 152463}
```

This template listens for the following state changed events:

Domain: media_player

My current media player definitions:
- 'Bose-ST10-1' (soundtouchplus integration)
- 'Bose-ST300' (soundtouchplus integration)
- 'SpotifyPlus Todd L' (spotifyplus integration)
- 'Spotify Todd L' (spotify integration)

**Comment by punxaphil** [comment]
Try https://github.com/punxaphil/custom-sonos-card/releases/tag/v7.8.0
It will show right of the volume bar

**Reply by thlucas1** [reply]
Thanks Phil, much appreciated.  I added it in the same place in my testing.  Lol

I am playing with the maxi media player now - will probably use it since it does not have the Sonos stuff embedded.  I have it on my dev environment now, and have made the following changes thus far:  added the power button in the volume.ts module; added options to show / hide all player control buttons vie the editor (no more css mods), and fixed what I think is a bug in the player mute logic.   Will send you a pull request for your consideration when I finish testing.   

Thanks for your help, and your efforts with the card.

**Reply by punxaphil** [reply]
> added the power button in the volume.ts module;

Already added:
https://github.com/punxaphil/maxi-media-player/releases/tag/v1.4.0

**Reply by thlucas1** [reply]
@punxaphil 
sent you a pull request.  It's been awhile since I created a pull request, so let me know if the format is to (or not) your liking.
Thanks - Todd

UPDATE 2024/04/09 09:45pm Central Time:
I just realized that we are still commenting in the custom-sonos-card discussion area.  The pull request I sent is for the maxi-media-player repo.  Sorry about that!

I also noticed that there is no `discussion` forum available on the maxi-media-player repo.  I would like to start a new discussion on how favorites work in the maxi-media-player card.  It seems like they only support the Sonos type of favorites, but I am not sure - I am not familiar with the Sonos favorites media browser format.  I have some ideas on how to let the user specify what kind of favorites they want to pull from the device; just need to know what the BrowseMedia dictionarys that are returned from the current function look like.  It appears it does a call to get the favorites root, then does repeated calls to get children items?  

Please let me know when you get a chance.  Thank you sir!  And please let me know if it's too much of an ask, or if I get too annoying.  I geek out on this type of stuff, and can get a little over-excited sometimes.  😄

**Reply by thlucas1** [reply]
Forgot to mention that I already had the power button changes finished when I started on the other items.  They may conflict with what you added, but if I remember correctly they did the same thing.   

Here's what I added, just to document it here:
- add editor settings to allow show / hide of player control buttons: repeat, shuffle, next track, previous track, and power.
- fixed an issue with media-control-service setVolumeMute and volumeDefaultStep to account for only 1 player; prior to this, it was ignoring the request (mute, volume up / down) if only one player was active.
- added power button to player-controls if current player state is off.
- when getting the speaker list, return the main player name if no pre-defined groups are defined.
